crates/binjs_io/src/entropy/predict.rs

```diff
@@ -8,128 +10,321 @@ use std::rc::Rc;
+        debug!(target: "entropy", "Converting ContextPredict {} to probabilities", description);
         let by_context = self.by_context.into_iter()
-            .map(|(context, by_key)| {
-                let instances : Vec<_> = by_key.values()
-                    .map(|x| *x as u32)
-                    .collect();
-                let distribution = Rc::new(RefCell::new(range_encoding::CumulativeDistributionFrequency::new(instances).
-                    unwrap()));
-                    // FIXME: This will fail if `by_key` is empty.
-                    // FIXME: We should have a fallback distribution in case everything is empty.
-
-                let by_key = by_key.into_iter()
-                    .enumerate()
-                    .map(|(index, (key, _))| {
-                        (key, Symbol {
-                            index,
-                            distribution: distribution.clone(),
-                        })
-                    })
-                    .collect();
-                // FIXME: Actually put the number of instances in `distribution`.
-                (context, by_key)
-            })
+            .map(|(context, info)| (context, info.instances_to_probabilities("ContextInformation")))
             .collect();
         ContextPredict {
-            by_context
+            by_context,
         }
     }
 }
 
-
-/// A predictor used to predict the probability of a symbol based on
-/// its position in a tree. The predictor is typically customized to
-/// limit the prediction depth, e.g. to 0 (don't use any context),
-/// 1 (use only the parent + child index) or 2 (use parent +
-/// grand-parent and both child indices).
-#[derive(Debug, Default, Serialize, Deserialize)]
-pub struct PathPredict<K, T> where K: Eq + Hash + Clone {
-    context_predict: ContextPredict<IOPath, K, T>,
+/// A specialized predictor used to predict possible values at a possible path in the AST.
+///
+/// This mechanism is meant to be used as follows:
+///
+/// 1. Create a new `PathPredict<NodeValue, usize>` and use `PathPredict::add` to count the number
+///     of instances of each possible `NodeValue` in each possible path. Alternatively, load this data
+///     from an existing dictionary.
+/// 2. Convert the `PathPredict<NodeValue, usize>` into a `PathPredict<NodeValue, SymbolInfo>`
+///     by calling `PathPredict::<_, usize>::instances_to_probabilities`.
+/// 3. Use method `PathPredict::<_, SymbolInfo>::by_node_value` and `by_node_value_mut` to get the statistics
+///     information in a specific path for a specific node value (used for compression). This information contains
+///     an index designed to be written to a compressed stream.
+/// 4. Use method `PathPredict::<_, SymbolInfo>::frequencies_at` to get the statistics information in a
+///     specific path for all node values that have shown up in this path (used for decompression). This
+///     information is used to extract an index from a compressed stream.
```

Comment:
does this mean that the index itself is not stored in the stream?

Reply:
Indeed, if we don't have `frequencies_at`, we don't have enough information in the stream to extract the index.